osutil/synctree.go

```diff
@@ -0,0 +1,138 @@
+	for relPath != "." {
+		if err := os.Remove(filepath.Join(baseDir, relPath)); err != nil {
+			// If the directory doesn't exist, then stop.
+			if os.IsNotExist(err) {
+				return nil
+			}
+			// If the directory is not empty, then stop.
+			if pathErr, ok := err.(*os.PathError); ok && pathErr.Err == syscall.ENOTEMPTY {
+				return nil
+			}
+			return err
+		}
+		relPath = filepath.Dir(relPath)
+	}
+	return nil
+}
+
+// EnsureTreeState ensures that a directory tree content matches expectations.
+//
+// EnsureTreeState walks subdirectories of the base directory, and
+// uses EnsureDirStateGlobs to synchronise content with the
+// corresponding entry in the content map.  Any non-existent
+// subdirectories in the content map will be created.
+//
+// After synchronising all subdirectories, any subdirectories where
+// files were removed that are now empty will itself be removed, plus
+// its parent directories up to but not including the base directory.
+//
+// No checks are performed to see whether subdirectories match the
+// passed globs, so it is the caller's responsibility to not create
+// directories that may match any globs passed in.
+//
+// For example, if the glob "snap.$SNAP_NAME.*" is used then the
+// caller should avoid trying to populate any directories matching
+// "snap.*".
+//
+// A list of changed and removed files is returned, as relative paths
+// to the base directory.
+func EnsureTreeState(baseDir string, globs []string, content map[string]map[string]*FileState) (changed, removed []string, err error) {
+	// Find all existing subdirectories under the base dir.  Don't
+	// perform any modifications here because, as it may confuse
+	// Walk().
+	subdirs := make(map[string]bool)
+	err = filepath.Walk(baseDir, func(path string, fileInfo os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !fileInfo.IsDir() {
+			return nil
+		}
+		relPath, err := filepath.Rel(baseDir, path)
+		if err != nil {
+			return err
+		}
+		subdirs[relPath] = true
+		return nil
+	})
+	if err != nil {
+		return changed, removed, err
+	}
+	// Ensure we process directories listed in content
+	for relPath := range content {
+		subdirs[relPath] = true
+	}
+
+	maybeEmpty := []string{}
+
+	// TODO: ensure that no directories in the tree match a
+	// directory glob (one that would match any globs that would
+	// be passed in a particular context).
+	for relPath := range subdirs {
+		dirContent := content[relPath]
+		path := filepath.Join(baseDir, relPath)
+		if err = os.MkdirAll(path, 0755); err != nil {
+			break
+		}
+		var dirChanged, dirRemoved []string
+		dirChanged, dirRemoved, err = EnsureDirStateGlobs(path, globs, dirContent)
+		changed = appendWithPrefix(changed, relPath, dirChanged)
+		removed = appendWithPrefix(removed, relPath, dirRemoved)
+		if err != nil {
+			break
+		}
+		if len(removed) != 0 {
+			maybeEmpty = append(maybeEmpty, relPath)
+		}
+	}
+	sort.Strings(changed)
```

Comment:
Small optimization hint: the return values of `EnsureDirStateGlobs` are sorted so perhaps we could iterate over subdirs in order and then simply append to a larger array that is in effect already sorted.

Because `filepath.Walk` visitor is called after internally sorting the list anyway, we could build a sorted list from get go, avoiding another explicit sorting operation.

Reply:
That doesn't quite give the same result as `sort.Strings`, since you'd end up with files before directories at a particular level.  It's still a stable ordering, so perhaps that would be okay.

What would become difficult is handling the error case: we'd need to reprocess subdirectories that had previously been successfully synchronised, possibly removing new files.  Perhaps we could perform a sort in this case, but now we've got a different stable ordering for the success and error cases :-(

Reply:
notice that we have places that assume they can use sort.SearchStrings   on the result of EnsureDirStateGlobs, so we need to be consistent around that, or give up on sorting and return sets (map[string]bool)  and make it the caller problem, across the lot

Reply:
Okay.  I'll leave the sorting as it currently is.